packages/devtools_app/lib/src/performance/performance_controller.dart

```diff
@@ -130,6 +123,43 @@ class PerformanceController
   /// This list is cleared and repopulated each time "Refresh" is clicked.
   List<TraceEventWrapper> allTraceEvents = [];
 
+  /// The tracking index for the first unprocessed trace event collected.
+  int nextTraceIndexToProcess = 0;
+
+  /// The tracking index for the first unprocessed [TimelineEvent] that needs to
+  /// be processed and added to the timeline events flame chart.
+  int nextTimelineEventIndexToProcess = 0;
+
+  /// Timeline events that have been processed.
+  ///
+  /// This may not include all of the trace events that have been collected.
+  /// Some may be unprocessed and sitting in [unprocessedTraceEvents].
+  ValueListenable<List<TimelineEvent>> get timelineEvents => _timelineEvents;
+  final _timelineEvents = ValueNotifier<List<TimelineEvent>>(<TimelineEvent>[]);
+
+  /// Whether flutter frames are currently being recorded.
+  ValueListenable<bool> get recordingFrames => _recordingFrames;
+  final _recordingFrames = ValueNotifier<bool>(true);
+
+  /// Frames that have been recorded but not shown because the flutter frame
+  /// recording has been paused.
+  final _pendingFlutterFrames = <FlutterFrame>[];
+
+  /// The collection of [TimelineEvent]s that should be linked to
+  /// [FlutterFrame]s but have not yet been assigned.
+  ///
+  /// These timeline events are keyed by the [FlutterFrame] ID specified in the
+  /// event arguments, which matches the ID for the corresponding
+  /// [FlutterFrame].
+  final unassignedFlutterFrameEvents = <int, List<TimelineEvent>>{};
```

Comment:
Maybe replace List<TimelineEvent> with a class with a clearer api.
Having the indexes in the List happen to be FlutterFrame ids is a little surprising.
class could be `UnasignedEvents` or similar.